### ISSUE TITLE
ECR architecture is not compatible with Epinio

### ISSUE THREAD

**Issue by pedro-brito-ilia** [epinio/epinio#1973]
On trying to integrate Epinio with ECR, found the following incompatibility:

ECR does not support automatic creation of new subfolders on it's repositories, meaning that for every new application deployed on Epinio we would need to manually create a new repository on ECR, which is not sustainable in the long run. I'm still trying to find a workaround for this, but these are my investigations so far:

For this use-case, I created a new ECR repository called test-ecr.
![image](https://user-images.githubusercontent.com/112405514/209171161-9ce02691-5d22-43cf-a83b-7e5a8dc78cc7.png)

Repository test-ecr only supports pushes with it's own name, here's a docker push for example:
![image](https://user-images.githubusercontent.com/112405514/209171579-30e41b24-fccd-4b29-a496-6731a30e2c50.png)

This is fine and makes sense, the problem is when we try to push a new image with a different name on that same repository, for example:
![image](https://user-images.githubusercontent.com/112405514/209171880-ab9c9e3b-2a92-4331-98c7-18a2e61a21dc.png)

In this new push, ECR waits undefinetly, until it finally throws out an EOF exception after some minutes. The only workaround I found for this is to manually create a new repository on ECR with this pattern, like this:
![image](https://user-images.githubusercontent.com/112405514/209172310-eaff2011-38fe-4ec7-abb5-b0fd80c8bc3e.png)

Now, when I try to push:
![image](https://user-images.githubusercontent.com/112405514/209172564-f0cf9b0a-7ccb-43d9-b7e2-323ac802ab39.png)

Now, in the way that Epinio handles images, it dinamically creates new ones for every application, this means that for every new deployment, we would need to manually create a new repository on ECR. This is not only a problem, but it's actually a blocker for integrating both these technologies.

For comparison, I'll show how JFrog Artifactory handles images when we try to push a new subfolder to a repository. I'll execute the same push, but now on our Artifactory URL:
![image](https://user-images.githubusercontent.com/112405514/209173430-91ec137a-4bd6-4ff0-b571-d2ea64bcc8e3.png)

The push worked imediatelly, without the need of manually creating a new repository, This is what Artifactory architecture looks like:
![image](https://user-images.githubusercontent.com/112405514/209173814-525f6fc6-06f1-4a7f-83ae-31bf0bb31221.png)

There's our subfolder! With the image and files inside of it, and this is repeatable for every new push. This is compatible with Epinio because every time a new application is deployed, the image with a new name is pushed and Artifactory automatically provides a subfolder.

In conclusion, we are trying to make our stack work with ECR, so I'm trying to find a workaround for this issue. Do any ideas come to your mind on how we could solve this?

**Comment by enrichman** [comment]
I faced the same issue while testing a bit the [IAM authentication](https://github.com/epinio/epinio/issues/1925#issuecomment-1363990134). I will have a look if there is a solution or workaround.

**Comment by enrichman** [comment]
Looking around I've found this blog post: https://www.paradigm.co/blog/automatic-aws-ecr-repository-creation

They have managed to fix this problem using a Lambda:

1. AWS CloudTrail logs an event when an attempt is made to push a docker image to a repository that doesn’t exist.
2. An AWS EventBridge rule triggers an AWS Lambda function on the above CloudTrail log event.
3. The AWS Lambda function makes an API call to create the missing repository.

Another possible solution could be leveraging the staging job that we have. We are using the aws cli to download the source code from S3. It could be possible to hack a bit the script, check if a repository with that name exists, and create it if not.
It's a bit of an hack/workaround but it should be possible.

The staging scripts are in a configmap, so it's even pretty easy to play with them: https://github.com/epinio/helm-charts/blob/main/chart/epinio/templates/stage-scripts.yaml

The `download` is using the aws cli: https://github.com/epinio/helm-charts/blob/3284ac51b321d9648b453658d33997d97fedc6ce/chart/epinio/templates/stage-scripts.yaml#L10-L25

**Comment by andreas-kupries** [comment]
For reference, ECR = [Elastic Container Registry](https://aws.amazon.com/ecr/).

**Comment by enrichman** [comment]
Hi @pedro-brito-ilia, I've tested the workaround and it seems to work. Could you try it and see if it could fit your case?

These are the steps to follow:

## Docker credentials

Probably you don't need this, since you already figured it out, but I'll write down a couple of commands just for reference.
Since the ECR doesn't support the IAM authentication we still need to configure the docker login.

To get the username and password we need to use the `aws ecr get-authorization-token` and decode the `authorizationToken`

```bash
$ aws ecr get-authorization-token | jq -r '.authorizationData[0].authorizationToken' | base64 -d
<USERNAME>:<PASSWORD>
```

We can export it in one command:
```bash
ECR_AUTH_TOKEN=$(aws ecr get-authorization-token | jq -r '.authorizationData[0].authorizationToken' | base64 -d) && \
    ECR_AUTH_TOKEN_USERNAME=$(echo $ECR_AUTH_TOKEN | cut -d':' -f1) && \
    ECR_AUTH_TOKEN_PASSWORD=$(echo $ECR_AUTH_TOKEN | cut -d':' -f2)
```

We also need the `registryNamespace` used

```
REGISTRY_NAMESPACE=$(kubectl get secret registry-creds -n epinio -o jsonpath='{.metadata.annotations.epinio\.io\/registry-namespace}')
```

Now we can delete the old docker credentials, create, and annotate the new one:

```bash
kubectl delete secret registry-creds --namespace epinio

kubectl create secret docker-registry registry-creds \
    --namespace epinio \
    --docker-server=<ECR_REGISTRY_ENDPOINT> \
    --docker-username=${ECR_AUTH_TOKEN_USERNAME} \
    --docker-password=${ECR_AUTH_TOKEN_PASSWORD}

kubectl annotate secret registry-creds --namespace epinio \
    epinio.io/registry-namespace=${REGISTRY_NAMESPACE} \
    kubed.appscode.com/sync='kubed-sync=registry-creds'

```

## IAM Role and policy

The aws cli in the staging job is using the `epinio-s3-connection-details` secret for the AWS configuration and credentials. The access key used should have the permission to create a ECR repository.

I will show you how to create and associate a policy to use a IAM role (if we are going to use the AWS S3 instead of Minio then the same policy with more permission could be used).

Create a policy with the permission to create ECR repositories

```bash
ECR_POLICY_NAME=EpinioECRPolicy

aws iam create-policy \
    --policy-name ${ECR_POLICY_NAME} \
    --policy-document \
'{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "ecr:CreateRepository"
            ],
            "Resource": [
                "*"
            ]
        }
    ]
}'
```

Attach it to the NodeInstanceRole of the cluster. To find it you can use the `kubectl -n kube-system describe configmap aws-auth` command:

```bash
NODEGROUP_ROLE_NAME=$(kubectl -n kube-system describe configmap aws-auth | grep rolearn | cut -d':' -f7 | cut -d'/' -f2)
AWS_ACCOUNT_ID=$(aws sts get-caller-identity | jq -r '.Account')

aws iam attach-role-policy \
    --role-name ${NODEGROUP_ROLE_NAME} \
    --policy-arn "arn:aws:iam::${AWS_ACCOUNT_ID}:policy/${ECR_POLICY_NAME}"
```


## Workaround

We are now ready to edit our script to dynamically create the repository.

Edit the `download` script in the `epinio-stage-scripts` configmap adding these lines:

```bash
echo '[profile ecr]\nregion=eu-central-1' > /tmp/aws_ecr_config
ECR_REPO=$(echo ${APPIMAGE} | cut -d'/' -f2- | cut -d':' -f1)
echo Creating ${ECR_REPO} repository
AWS_CONFIG_FILE=/tmp/aws_ecr_config aws --profile ecr ecr create-repository --repository-name ${ECR_REPO}
```

This will create an empty AWS config file, that will be used during the `aws ecr create-repository` command. Since the credential is empty then the IAM role of the EC2 instance will be used, and so the policy that we created will apply.

Let me know if this works for you.

**Comment by pedro-brito-ilia** [comment]
Hi @enrichman, after further testing this workaround, I got the following issue when trying to push an app:

Error: ErrImagePull
Error: ImagePullBackOff
Pulling image "ghcr.io/epinio/epinio-unpacker:1.0"
Failed to pull image "ghcr.io/epinio/epinio-unpacker:1.0": rpc error: code = Unknown desc = manifest unknown
Back-off pulling image "ghcr.io/epinio/epinio-unpacker:1.0"

Maybe this is related to the removal of Epinio default Docker registry configuration? Should I pull this image to my external registry first?

**Comment by pedro-brito-ilia** [comment]
@enrichman our team was able to solve this. It was a general issue on 1.5.1 version using epinio-unpacker:1.0, upgrading to 1.5.2 solved it.

Besides that, your workaround worked like a charm :) Just tested it using a sample conda app, I'm going to run more tests in the next days using more complex and secure applications, but I think we can close this issue for now.
Maybe we could discuss including these required changes on a next version, so this is consistent with upgrades and restarts.

**Comment by andreas-kupries** [comment]
@pedro-brito-ilia 
> our team was able to solve this. It was a general issue on 1.5.1 version using epinio-unpacker:1.0, upgrading to 1.5.2 solved it.

Can you tell us what the issue was with `epinio-unpacker:1.0` ?

**Comment by enrichman** [comment]
Thanks, I will close this issue then.

**Comment by pedro-brito-ilia** [comment]
@andreas-kupries sure. epinio-unpacker version is set to 1.0 on Epinio 1.5.1, and apparently this image does not exist anymore. Upgrading Epinio solved it as in 1.5.2 the version is not set, so Epinio just pulled the latest epinio-unpacker image.

**Comment by andreas-kupries** [comment]
Ouch about the image for 1.0 being gone. Not sure where that happened. Will have to bring that up with the team.
We should not remove images which are referenced by a release.

Nitpicking: When the version is not set it actually does not pull the "latest", but rather the version of the chart, i.e. "1.5.2".
Same setup as for the `epinio-server` image.

**Comment by pedro-brito-ilia** [comment]
Good to know :) Thanks!
